tracer/src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs

```diff
@@ -326,18 +169,69 @@ private static Span StartMvcCoreSpan(Tracer tracer, Span parentSpan, BeforeActio
             ActionDescriptor actionDescriptor = typedArg.ActionDescriptor;
             IDictionary<string, string> routeValues = actionDescriptor.RouteValues;
 
-            string controllerName = routeValues.TryGetValue("controller", out controllerName)
-                                        ? controllerName?.ToLowerInvariant()
-                                        : null;
-            string actionName = routeValues.TryGetValue("action", out actionName)
-                                    ? actionName?.ToLowerInvariant()
-                                    : null;
-            string areaName = routeValues.TryGetValue("area", out areaName)
-                                  ? areaName?.ToLowerInvariant()
-                                  : null;
-            string pagePath = routeValues.TryGetValue("page", out pagePath)
-                                  ? pagePath?.ToLowerInvariant()
-                                  : null;
+            string controllerName = null;
+            string actionName = null;
+            string areaName = null;
+            string pagePath = null;
+
+            // if RouteValues is a dictionary we can cast and use the internal Enumerator with no allocations.
+            if (routeValues is Dictionary<string, string> dctRouteValues)
+            {
+                // We can also avoid calculating the key hash all the time and just enumerate and bailout when we have all the data.
+                int finishBits = 0;
+                foreach (var routeValue in dctRouteValues)
+                {
+                    // 1 | 2 | 4 | 6 Flags
+                    if (finishBits == 13)
+                    {
+                        break;
+                    }
+
+                    if (routeValue.Key == "controller")
+                    {
+                        controllerName = routeValue.Value;
```

Comment:
I think these need to be like this?
```suggestion
                        controllerName = routeValue.Value?.ToLowerInvariant();
```

Reply:
done in https://github.com/DataDog/dd-trace-dotnet/pull/4390/commits/650371eb9f01ba9d28929c87761a48f80b46d8d0